# Zebra Finch SNP calling 

This document describes the mapping and SNP calling from paired end Illumina sequence data from 10 zebra finch
individuals sampled from Australia. The birds were originally sequenced and mapped to the zebra finch genome in the study of [Singhal et al. (2015)](https://www.ncbi.nlm.nih.gov/pubmed/26586757).

## Public Availibility of sequence data and files
The BAM files used below are publically available from [here](http://www.ebi.ac.uk/ena/data/view/PRJEB10586)

## Programs used in this pipeline
__Versions of programs used__

* GATK 3.4-46-gbc02625
* samtools 1.2
* bcftools 1.2
* freebayes 1.02
* faList from https://www.biostat.wisc.edu/~cdewey/software/cndsrc-2013.01.11.tar.gz

## Steps in the pipeline

### Prepare the reference genome files
Excluded the chrZ, chrZ_random and mitochondrial genome from snp calling.

    $ python reorder_reference.py /data/bo1pgc/zebra_finch_genome/taeGut1.fa > /data/bo1pgc/zebra_finch_genome/taeGut1.reordered.fa
    $ faList < /data/bo1pgc/zebra_finch_genome/taeGut1.reordered.fa | grep -v chrM | grep -v chrZ > chromosomes.txt
    $ qsub prepare_reference.sh

### Prepare the downloadeded BAM files for the 10 birds
    
    $ qsub index_bam_files.sh

### Generate the gVCF files for each individual using GATK's HaplotypeCaller

    $ sh haplotype_caller.sh bamlist.list chromosomes.txt

### Run the GenotypeGVCF program to create all-sites VCF files 

    $ ls /fastdata/bo1pgc/parus_reseq/singhal_zebra_finch_data/zf_snp_calling/gVCF/*.g.vcf.gz > gvcflist.list
    $ sh genotype_gvcfs.sh gvcflist.list chromosomes.txt zf_10birds
        
### Variant Quality Score Recalibration

#### Generating the training set

Called SNPs with Freebayes using the following command:
    
    $ freebayes -f  -L bamlist.list  --report-monomorphic --report-genotype-likelihood-max --min-mapping-quality 20 --min-base-quality 10
    
Extracted and filtered SNPs based on depth and Quality

	$ python extract_SNPS_only.py
	$ qsub sort_and_bgzip.sh

Intersected the GATK SNPs with the filtered Freebayes SNPs to generate the training set using bcftools isec program.

    $ bcftools isec -p /fastdata/bo1tg/zebra_VCF_10birds/VCF_merge/Merged_VCF /fastdata/bo1tg/zebra_VCF_10birds/VCF_merge/zf_10birds.gatk.SNPsONLY.sorted.vcf.gz /fastdata/bo1tg/zebra_VCF_10birds/VCF_merge/zf_10birds.freebayes.SNPsONLY.sorted.vcf.gz -n=2 -w 1

#### Running and applying VQSR
Used training set to run VQSR and applied the recalibration tables generated by the VariantRecalibrator tool in GATK. 

    $ qsub vqsr.sh
    $ python apply_recalibration.py
    
###  Repetitive Region filter

Marked sites in the recalibrated VCF files that were located in repetitive regions of the genome.

    $ qsub filter_repeats.zf.sh /fastdata/bo1tg/VCF_Backup_Padraic/zebra_finch/All_sites/recalibrated_snps_99.vcf.gz /fastdata/bo1tg/VCF_Backup_Padraic/zebra_finch/All_sites/recalibrated_snps_99.rep_filtered.vcf.gz 
